[Tang Zhi You] iP

Zhi-You opened the pull request.

Zhi-You committed:
Add initial skeletal version that greets the user and exits

Zhi-You committed:
Add comments for level-0

Zhi-You committed:
Add input and echo functionalities

Zhi-You committed:
Add storing of tasks and displaying of tasks capabilities

Zhi-You committed:
Add Task class to represent tasks

Zhi-You committed:
Add mark task as done functionality

Zhi-You committed:
Tweak line length to be no longer than 120 chars by using line wrapping

Zhi-You committed:
Add DisplayManager class to handle formatting of output to user

Zhi-You committed:
Make taskCount into a class-level member in Task class

Zhi-You committed:
Add subclasses Todo, Deadline and Event

Zhi-You committed:
Add support for tracking three types of tasks

Users can now input three type of tasks - Todo, Deadlines and Events
Duke will distinguish what type of tasks are being added to the list
Duke will also perform appropriate actions for the different commands
related to the tasks

Code is yet to be refactored

Zhi-You committed:
Make methods in DisplayManager into class-level methods to allow
for use in Duke without instantiating an object

Zhi-You committed:
Add test using I/O redirection technique capability to semi-automate
the testing of Duke

Zhi-You committed:
Remove magic literals

Zhi-You committed:
Improve code quality mainly by refactoring code

Followed closely to 'Code Quality' topics. Followed SLAP.

Zhi-You committed:
Move main to top of Duke.java file for logical flow

Zhi-You committed:
Move printing of Duke logo and hello message to DisplayManager

Part of refactoring process

mrwsy1 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+import java.util.Scanner;
+
 public class Duke {
```
Nice code, very neat.

mrwsy1 reviewed (commented):
Looks good.

Zhi-You committed:
Make tick and cross symbols into constants to avoid having magic
literals

Zhi-You committed:
Add DukeException to represent exceptions specific to Duke

Add enum ErrorType to specify the different error types

Zhi-You committed:
Add ErrorTypeManager and overall error differentiation capabilities

Zhi-You committed:
Add todo empty description error handling

Zhi-You committed:
Add error handling for event, mark task and deadline

Handle exceptions caused by empty input and wrong formatting
Make scanner global so that we dont keep initializing it in the
while loop which causes a bug whereby the first scanner will take away
the next line before the second scanner can

Zhi-You committed:
Merge branch 'branch-Level-5'

Zhi-You committed:
Modify automated test to test error handling too

Zhi-You committed:
Organize classes into duke and task classes into task

Zhi-You committed:
Merge branch 'branch-A-Packages'

Zhi-You committed:
Tweaks to touch up on week 5 work before working on week 6 work

Zhi-You committed:
Revert "Tweaks to touch up on week 5 work before working on week 6 work"

This reverts commit 3ad498da7024453554e2ad1173abb1f2c0513407.

Zhi-You committed:
Tweak to week 5 work before starting on week 6 work

Zhi-You committed:
Use ArrayList<Task> to store tasks

Zhi-You committed:
Add support for deleting tasks from the list.

Zhi-You committed:
Add save and load tasks functionalities

Save the tasks in the hard disk automatically whenever the task list
changes. Load the data from the hard disk when Duke starts up.

Zhi-You committed:
Add error handling for save and load related functions

Zhi-You committed:
Remove magic literals from deadline and event delimiters

Zhi-You committed:
Merge branch 'branch-Level-6'

Zhi-You committed:
Merge branch 'branch-Level-7' and resolve conflicts to achieve both
delete and save and load functionalities

# Conflicts:
#	src/main/java/duke/DisplayManager.java
#	src/main/java/duke/Duke.java
#	src/main/java/duke/ErrorTypeManager.java

Zhi-You committed:
Add data folder into gitignore

Zhi-You committed:
Add data folder and automated test's data folder into gitignore

Zhi-You committed:
Move constants positions for logical flow

Zhi-You committed:
Package App as a JAR file

Zhi-You committed:
Fix .gitignore

Bencotti commented on src/main/java/duke/DataManager.java:
```diff
@@ -0,0 +1,156 @@
+    // Specifying file paths in an OS-independent way
+    private static final String CURRENT_DIRECTORY = System.getProperty("user.dir");
+    private static final Path DATA_DIRECTORY = Paths.get(CURRENT_DIRECTORY, "data");
+    private static final Path DATA_PATH = Paths.get(CURRENT_DIRECTORY, "data", "duke.txt");
```
Thoughtful implementation! Remember to check if it works on multiple OS just to be safe!

Bencotti commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,283 @@
+            DisplayManager.printFileErrorMessage();
+        }
+
+        boolean exitCommandPassed = false;
```
Consider renaming your boolean variable to sound more like a boolean, at the moment it sounds more like a statement

Bencotti commented on src/main/java/duke/DataManager.java:
```diff
@@ -0,0 +1,156 @@
+        tasks.add(new Todo(taskDescription));
+
+        // Recover its isDone status
+        markTaskDoneIfDone(task, tasks);
```
The method name `markTaskDoneIfDone` might be confusing since there exists a `markAsDone`, to improve readability of code perhaps you could consider renaming it?

Bencotti commented on src/main/java/duke/DataManager.java:
```diff
@@ -0,0 +1,156 @@
+        String[] deadlineTaskParts = taskDescription.split(DEADLINE_TASK_DELIMITER);
+        String deadlineTaskDescription = deadlineTaskParts[0];
+        // Took the substring of the deadline portion because we want to ignore the closing bracket at the end
+        String deadline = deadlineTaskParts[1].substring(0, deadlineTaskParts[1].length() - 1);
```
Since the magic number "1" is used in multiple places,  assigning it to a variable to explain what it does would value add to the developer and new devs by making its intention clear

Bencotti reviewed (commented):
Great work! Could see the effort you put in following the coding standards and implementing the code readability principles taught in class. 👍

Zhi-You committed:
Add TaskList class to contain list of tasks

TaskList has the operations to add, delete, mark as done and list tasks

Zhi-You committed:
Extract Command classes.

Each command is its own class now and they have methods for their own
purposes.

Command is the parent class, and it is an abstract class.

Each Command class has the execute function which mainly control their
behaviour in the program.

Zhi-You committed:
Add Parser class to parse commands and task description

Tweak other classes and methods to work with the newly added Command
classes and parser.

Refactor code to make it more OOP

Zhi-You committed:
Refactor closely related classes into packages

Zhi-You committed:
Tweak test cases

Zhi-You committed:
Tweak code to follow coding quality

Zhi-You committed:
Tweak welcome message

Zhi-You committed:
Add understanding dates capabilities with DateParser class and methods

Zhi-You committed:
Add storing of date in deadline function

Zhi-You committed:
Add storing date and updating eventTime functionalities for Event tasks

Zhi-You committed:
Add automated test cases for date functionalities

Zhi-You committed:
Add FindCommand

Zhi-You committed:
Add find tasks capabilities by matching description with keyword.

Streams were used to easily filter the tasks that match the keyword
in their task description.

Zhi-You committed:
Add test cases for find function for automated testing

Zhi-You committed:
Add javadoc comments to Command, DeadlineCommand and DeleteCommand

Zhi-You committed:
Add JavaDoc comments to all Commands

Zhi-You committed:
Add JavaDoc comments and comments for exception classes and parser

Zhi-You committed:
Add JavaDoc comments and comments for all Task types and TaskList.

Zhi-You committed:
Add JavaDoc comments and comments to Ui and Duke.

Zhi-You committed:
Merge pull request #1 from Zhi-You/branch-Level-8

Branch level 8

Zhi-You committed:
Merge branch 'master' into branch-Level-9
Fix merge conflicts in EXPECTED.TXT and input.txt

# Conflicts:
# text-ui-test/EXPECTED.TXT
# text-ui-test/input.txt

Zhi-You committed:
Merge pull request #2 from Zhi-You/branch-Level-9

Branch level 9

Zhi-You committed:
Merge branch 'master' into branch-A-JavaDoc
Fix merge conflicts and ready to merge pull request.

# Conflicts:
#	src/main/java/duke/exception/ErrorTypeManager.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/ui/Ui.java

Zhi-You committed:
Add JavaDoc comments to classes related to Find functionalities

Zhi-You committed:
Merge pull request #3 from Zhi-You/branch-A-JavaDoc

Branch a java doc

Zhi-You committed:
Set theme jekyll-theme-cayman

Zhi-You committed:
Update README.md

Zhi-You committed:
Update README.md as a completed user guide.

Zhi-You committed:
Add bye feature  to README.md

Zhi-You committed:
Update README.md to make it more concise and clear

Zhi-You committed:
Add format for usage

Zhi-You committed:
Set theme jekyll-theme-slate

Zhi-You committed:
Update _config.yml

Zhi-You committed:
Set theme jekyll-theme-cayman

Zhi-You committed:
Add link to Github pages onto Header of README.md
